### PR TITLE
[tutor] Add AI-Powered Installation Tutor

### PR DESCRIPTION
**Pull request by srikrishnavansi**
---
  ## Summary
  Add comprehensive AI-powered tutor for package education with:
  - Interactive tutoring using TutorAgent orchestration
  - LLM-powered lessons, code examples, and Q&A (Claude API via cortex.llm_router)
  - SQLite-based progress tracking and student profiling
  - CLI integration via `cortex tutor <package>`

  ## Related Issue
  Closes cortexlinux/cortex-distro#30

  ## Type of Change
  - [x] New feature

  ## Demo

  https://github.com/user-attachments/assets/643526af-1252-4436-b758-250211fb3159

  ## AI Disclosure
  - [x] AI/IDE/Agents used (please describe below)

  Used **Claude (Anthropic)** via Claude Code CLI for:
  - Code implementation and architecture design
  - PR review fixes (addressing Gemini/CodeRabbit comments)
  - Documentation generation

  ## Testing
  - Run: `pytest tests/tutor/ -v`
  - Manual testing: `cortex tutor docker --fresh`

  ## Checklist
  - [x] Tests pass locally
  - [x] Code follows style guide
  - [x] Documentation updated (AI_TUTOR.md)
  - [x] CLA signed (PR cortexlinux/cortex#570)

## Summary by CodeRabbit

* **New Features**
  * Added an AI-powered interactive tutor CLI: list packages, view/reset progress, ask questions, or start interactive lessons with progress tracking, branded terminal UI, lesson generation, deterministic fallbacks, and persistent progress storage.

* **Documentation**
  * Added comprehensive AI tutor user and developer documentation.

* **Tests**
  * Added extensive unit and integration tests covering CLI, agent, LLM flows, tools, storage, validators, and interactive sessions.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>

### CODE REVIEW

**Review by Anshgrover23 — CHANGES_REQUESTED** [review]
@srikrishnavansi Your SonarQubeCloud Quality Gate is failing kindly address that issues. Also address coderabbit comments that are still open.

File: requirements.txt [hunk 1]
```diff
@@ -23,3 +23,9 @@ pyyaml>=6.0.0
 # Type hints for older Python versions
 typing-extensions>=4.0.0
 PyYAML==6.0.3
+
+# AI Tutor Dependencies (Issue #131)
+langchain>=0.3.0,<2.0.0
```
Comment (Anshgrover23) [inline requirements.txt]: Why need of a langchain dependency, I think we can do without this.

**Review by Anshgrover23 — CHANGES_REQUESTED** [review]

**Comment by srikrishnavansi** [comment]
@coderabbitai

**Comment by srikrishnavansi** [comment]
@coderabbitai review

**Comment by Anshgrover23** [comment]
@srikrishnavansi First rule to contribute in open source: Be Respectful to maintainers.
Also, again Purely AI Slop Code.

**Comment by srikrishnavansi** [comment]
Hi, thanks for taking the time to review.

  I'm here to learn and contribute. I spent over a day on this outside my regular work hours - studying your codebase, writing tests, fixing CI issues, and iterating on feedback. I understand it didn't meet your expectations, and that's okay.

  But "AI slop" doesn't help me understand what to improve. Was it the architecture? The coding style? The feature itself? A few specific pointers would go a long way for someone genuinely trying to contribute.

  Either way, I appreciate you maintaining this project. No hard feelings, and good luck with it.

**Comment by Anshgrover23** [comment]
>I'm here to learn and contribute. I spent over a day on this outside my regular work hours - studying your codebase, writing tests, fixing CI issues, and iterating on feedback. I understand it didn't meet your expectations, and that's okay.

Thanks for the effort and the working demo.

* This is about scope and architectural fit, not effort.
* The PR is too large 10k lines of changes, including ~3.3k lines of tests and ~1.5k lines of documentation,
* Introduces heavy dependencies.
* Many tests are shallow and inflate the reviewer burden.

>But "AI slop" doesn't help me understand what to improve. Was it the architecture? The coding style? The feature itself? A few specific pointers would go a long way for someone genuinely trying to contribute.

When I used the term “AI slop”, I was referring to unchecked, low-value generated code and tests that inflate the diff and reviewer burden, not denying that you put thought into parts of the design. As a maintainer, I have to judge the PR as a whole, not individual files in isolation.






> 
> Either way, I appreciate you maintaining this project. No hard feelings, and good luck with it.

* This PR is closed because it’s oversized and misaligned.
* For future contributions: keep PRs small and focused, reuse existing `cortex/llm`, avoid new frameworks without prior discussion, and add tests that assert real behavior.

**Comment by srikrishnavansi** [comment]
@coderabbitai review

File: cortex/tutor/__init__.py [hunk 2]
```diff
@@ -0,0 +1,13 @@
+"""
+Intelligent Tutor - AI-Powered Installation Tutor for Cortex Linux.
+
+An interactive AI tutor that teaches users about packages and best practices.
+"""
+
+__version__ = "0.1.0"
+__author__ = "Sri Krishna Vamsi"
```
Comment (Anshgrover23) [inline cortex/tutor/__init__.py]: remove this

**Review by Anshgrover23 — CHANGES_REQUESTED** [review]
@srikrishnavansi The current structure has 7+ `__init__.py` files for what could be 10 Python files in one folder.


**Current (over-engineered):**
```
tutor/
├── agents/
│   ├── __init__.py
│   └── tutor_agent/
│       ├── __init__.py
│       └── tutor_agent.py
├── contracts/
│   ├── __init__.py
│   ├── lesson_context.py
│   └── progress_context.py
├── memory/
│   ├── __init__.py
│   └── sqlite_store.py
├── prompts/                    # markdown files, fine
├── tools/
│   ├── __init__.py
│   └── deterministic/
│       ├── __init__.py
│       ├── lesson_loader.py
│       ├── progress_tracker.py
│       └── validators.py
├── __init__.py
├── branding.py
├── cli.py
├── config.py
└── llm.py
```



The whole tutor module could be simplified:

**Better (flattened):**
```
tutor/
├── prompts/                    # keep as subfolder for markdown
├── __init__.py
├── agent.py                    # was agents/tutor_agent/tutor_agent.py
├── branding.py
├── cli.py
├── config.py
├── contracts.py                # merge lesson_context + progress_context
├── llm.py
├── sqlite_store.py             # was memory/sqlite_store.py
├── tools.py                    # merge all deterministic tools
└── validators.py               # was tools/deterministic/validators.py
```

**Comment by Anshgrover23** [comment]
@srikrishnavansi


The following files appear to be completely unused. I could not find any imports, runtime references, or loading logic for them anywhere in the codebase.

Please remove these to avoid carrying dead code.

| Path | Status |
|---|---|
| `cortex/tutor/prompts/` (entire directory) | Unreferenced |
| `cortex/tutor/prompts/agents/tutor/system.md` | Unused |
| `cortex/tutor/prompts/tools/examples_provider.md` | Unused |
| `cortex/tutor/prompts/tools/lesson_generator.md` | Unused |
| `cortex/tutor/prompts/tools/qa_handler.md` | Unused |

**Comment by srikrishnavansi** [comment]
Summary of Changes

  1. Structured Outputs - Refactored llm.py to use Anthropic's tool_use feature with Pydantic models for structured LLM outputs instead of manual JSON parsing
  2. New Response Models - Added LessonResponse, QAResponse, and QACodeExample models in contracts.py for type-safe structured outputs
  3. Dead Code Removal - Removed unused classes (LessonPlanOutput, LessonReflectionOutput) and constants (KNOWN_PACKAGE_CATEGORIES)
  4. Input Validation - Added validate_question() call in cli.py before processing user questions
  5. SonarQube Fixes - Replaced conlist with plain list[T] types to resolve unexpected argument warnings
  6. Test Updates - Rewrote test_llm.py to match new structured output implementation with proper mocking
  7. Documentation - Updated AI_TUTOR.md to reflect the new structured outputs architecture


  @Anshgrover23 Please let me know if any further changes are needed!

File: cortex/tutor/cli.py [hunk 3]
```diff
@@ -0,0 +1,390 @@
...
+        console.print()
+        tutor_print("Session ended", "info")
+        return 0
+    except Exception as e:
+        print_error_panel(f"An error occurred: {e}")
+        if verbose:
+            import traceback
+
+            console.print_exception()
+        return 1
```
[the remark this answers is not in the document]
Comment (Anshgrover23) [reply cortex/tutor/cli.py]: @srikrishnavansi Please address this one.

[the remark this answers is not in the document]
Comment (srikrishnavansi) [reply cortex/tutor/cli.py]: @Anshgrover23  ,This issue has already been addressed in the current code. The fresh parameter is passed to InteractiveTutor at line 123 (InteractiveTutor(package, force_fresh=fresh)), and InteractiveTutor.__init__ accepts this parameter (def __init__(self, package_name: str, force_fresh: bool = False)). CodeRabbit may be reviewing an older version of the code.

**Review by mikejmorgan-ai — DISMISSED** [review]
## Review: APPROVED ✅

**AI-Powered Installation Tutor (PR #566)**

### Code Quality
- ✅ Clean architecture: agent.py, cli.py, config.py, validators.py, sqlite_store.py, tools.py
- ✅ Strong input validation with blocked dangerous patterns (rm -rf, fork bombs, eval injection)
- ✅ Secure file permissions (0o700 for data directory)
- ✅ Proper error handling with user-friendly messages
- ✅ Lazy imports for LLM components (only loaded when API key available)

### Security
- ✅ BLOCKED_PATTERNS prevents command injection attempts
- ✅ Package name validation: alphanumeric + dots/hyphens/underscores only
- ✅ Input length limits prevent abuse
- ✅ Sanitization removes null bytes and truncates long inputs
- ✅ API key validated before use

### Test Coverage  
- ✅ 167 tests all passing
- ✅ Tests cover: CLI, tools, validators, progress tracking, LLM integration
- ✅ Edge cases: blocked patterns, empty inputs, invalid scores, etc.

### Architecture
- ✅ `TutorAgent` orchestrates lesson generation and Q&A
- ✅ `InteractiveTutor` provides menu-driven CLI experience
- ✅ `SQLiteStore` handles progress persistence
- ✅ `LessonLoaderTool` manages caching with fallback templates

### Integration
- ✅ Clean integration with main CLI via `cortex tutor <package>`
- ✅ Supports: --list, --progress, --reset, --fresh, -q flags
- ✅ Documentation in docs/AI_TUTOR.md

### Notes
- Requires Python 3.10+ (union type hints)
- Requires ANTHROPIC_API_KEY for LLM features
- Offline fallback lessons available for common packages (docker, git, nginx)

Ready for merge once Ansh approves.

Reviewed-by: Claude Code

File: cortex/tutor/llm.py [hunk 4]
```diff
@@ -0,0 +1,258 @@
...
+    Call Anthropic API with structured output using tool use.
+
+    Returns:
+        Tuple of (parsed response model, cost in USD)
+    """
+    client = get_client()
+    tool = _create_tool_from_model(tool_name, tool_description, response_model)
+
+    response = client.messages.create(
+        model="claude-sonnet-4-20250514",
```
Comment (Anshgrover23) [inline cortex/tutor/llm.py]: Do not hardcode model name, instead should be configurable via Config or environment variable

File: cortex/tutor/llm.py [hunk 5]
```diff
@@ -0,0 +1,258 @@
...
+import os
+from typing import TypeVar
+
+import anthropic
+from pydantic import BaseModel, ValidationError
+
+from cortex.tutor.contracts import LessonResponse, QAResponse
+
+logger = logging.getLogger(__name__)
+_client: anthropic.Anthropic | None = None
```
Comment (Anshgrover23) [inline cortex/tutor/llm.py]: There is no proper thread-safe initialization (no locks), for e.g  `SQLiteStore` uses `RLock`.

File: cortex/tutor/llm.py [hunk 6]
```diff
@@ -0,0 +1,258 @@
...
+        max_tokens=max_tokens,
+        temperature=temperature,
+        system=system_prompt,
+        tools=[tool],
+        tool_choice={"type": "tool", "name": tool_name},
+        messages=[{"role": "user", "content": user_prompt}],
+    )
+
+    # Extract tool use result
+    for block in response.content:
```
Comment (Anshgrover23) [inline cortex/tutor/llm.py]: What if `response.content` is empty or None ?

File: cortex/tutor/branding.py [hunk 7]
```diff
@@ -0,0 +1,266 @@
...
+    try:
+        response = console.input(prompt_text)
+        return response.strip() or default or ""
+    except (EOFError, KeyboardInterrupt):
+        console.print()
+        tutor_print("Operation cancelled", "info")
+        return ""
+
+
+def create_progress_bar(_description: str = "Processing") -> Progress:
```
Comment (Anshgrover23) [inline cortex/tutor/branding.py]: Remove unused `_description`.

File: cortex/tutor/config.py [hunk 8]
```diff
@@ -0,0 +1,143 @@
...
+from pathlib import Path
+
+from dotenv import load_dotenv
+from pydantic import BaseModel, Field, field_validator
+
+# Load environment variables from .env file
+load_dotenv()
+
+# Default number of topics for progress tracking
+DEFAULT_TUTOR_TOPICS = 5
```
Comment (Anshgrover23) [inline cortex/tutor/config.py]: The fallback to 5 topics seems arbitrary. What if a package has more/fewer topics? Also, do not use magic numbers.

**Comment by srikrishnavansi** [comment]
@Anshgrover23 

  Code Changes:

  1. Hardcoded model name - Now configurable via TUTOR_MODEL_NAME env variable (default: claude-sonnet-4-20250514)
  2. No thread-safe client initialization - Added threading.Lock with double-checked locking pattern in llm.py
  3. Empty/None response.content not handled - Added null check: if not response.content: raise ValueError(...)
  4. Duplicate package validation in cli.py - Removed. agent.py handles validation, CLI catches ValueError
  5. Unused _description parameter - Removed from create_progress_bar()
  6. Error response structure inconsistent - Added _error_response() helper for consistent error structure
  7. Magic number DEFAULT_TUTOR_TOPICS = 5 - Renamed to DEFAULT_TUTOR_TOPICS_COUNT with clear purpose
  8. Cache TTL not configurable - Now configurable via TUTOR_CACHE_TTL_HOURS env variable (default: 24)
  9. clear_cache docstring misleading - Updated docstring to clarify behavior
  10. --fresh flag not propagated - Already addressed. InteractiveTutor accepts force_fresh parameter

  Documentation Updates:

  Updated docs/AI_TUTOR.md with new environment variables (TUTOR_MODEL_NAME, TUTOR_CACHE_TTL_HOURS)

  Ready for re-review!

**Review by Anshgrover23 — CHANGES_REQUESTED** [review]
@srikrishnavansi There are still comments left above opened.

**Review by Anshgrover23 — CHANGES_REQUESTED** [review]
@srikrishnavansi Why are your comments are also completely AI-Generated, kindly write in your own human tone.
Also, haven't addressed this one:
image

File: cortex/tutor/tools.py [hunk 9]
```diff
@@ -0,0 +1,479 @@
...
+                "error": str(e),
+            }
+
+    def cache_lesson(
+        self,
+        package_name: str,
+        lesson: dict[str, Any],
+        ttl_hours: int = 24,
+    ) -> bool:
+        """Cache a lesson for future retrieval."""
```
[the remark this answers is not in the document]
Comment (srikrishnavansi) [reply cortex/tutor/tools.py]: Already Fixed.
